src/agent/coverage/src/block/linux.rs

```diff
@@ -0,0 +1,512 @@
+            format_err!("no loadable segments for ELF object ({})", module.display())
+        })?;
+
+    let mut blocks = vec![];
+
+    for sym in elf.symbols() {
+        if sym.kind() == object::SymbolKind::Text && sym.size() > 0 {
+            if let Some(idx) = sym.section().index() {
+                let section = elf.section_by_index(idx)?;
+                let sym_blocks = find_symbol_blocks(section, sym)?;
+                blocks.extend(sym_blocks);
+            }
+        }
+    }
+
+    for block in &mut blocks {
+        block.offset -= load_va;
+    }
+
+    Ok(blocks)
+}
+
+pub fn find_symbol_blocks(section: ElfSection, sym: ElfSymbol) -> Result<Vec<Block>> {
+    // Slice symbol's data within section.
+    let lo = (sym.address() - section.address()) as usize;
+    let hi = lo + (sym.size() as usize);
+    let data = section.data()?;
+    let data = &data[lo..hi];
+
+    // Find basic block leaders, as VAs relative to the section load addr.
+    let leaders = find_function_leaders(data, sym);
+
+    let mut decoder = Decoder::new(64, data, DecoderOptions::NONE);
+
+    // Enable using `Decoder::ip()` to get module-relative offsets.
+    decoder.set_ip(sym.address());
+
+    let mut blocks = vec![];
+
+    let mut offset = 0;
+    let mut size = 0;
+
+    let mut inst = Instruction::default();
+    while decoder.can_decode() {
+        let current = decoder.ip() as u64;
+
+        decoder.decode_out(&mut inst);
+
+        let is_leader = leaders.contains(&current);
+
+        // May not point to an actual instruction in the function span.
+        let next = decoder.ip() as u64;
+
+        // If the next instruction is a leader, or the end of the function, then
+        // we have reached the end of the block.
+        let end = !decoder.can_decode() || leaders.contains(&next);
+
+        size += inst.len() as u64;
+
+        if is_leader {
+            offset = current;
+        }
+
+        // Note: for a 1-instruction block, `is_leader && end`.
+        if end {
+            blocks.push(Block { offset, size });
+
+            // Reset WIP block.
+            offset = 0;
+            size = 0;
+        }
+    }
+
+    Ok(blocks)
+}
+
+/// Compute the basic block leaders of a function as virtual addresses.
+fn find_function_leaders(data: &[u8], sym: ElfSymbol) -> BTreeSet<u64> {
+    let mut decoder = Decoder::new(64, data, DecoderOptions::NONE);
+    decoder.set_ip(sym.address());
+
+    // Contains leaders with VAs, assuming section load address.
+    let mut leaders = BTreeSet::new();
+
+    // Function entry is a leader.
+    leaders.insert(sym.address());
+
+    let mut inst = Instruction::default();
+    while decoder.can_decode() {
+        decoder.decode_out(&mut inst);
+
+        if let Some(target) = branch_target(&inst) {
+            // The branch target is a leader.
+            leaders.insert(target);
+
+            // The next instruction is a leader, if it exists.
+            if decoder.can_decode() {
```

Comment:
The Windows and Linux code should be unified.

This condition points out to me distinct (but probably harmless) bugs in both platforms:

* Linux (here) treats the fall thru instruction of an unconditional branch as a leader which isn't necessarily correct.
* The very similar Windows code incorrectly assumes the fall thru instruction of a conditional branch is reachable.

Reply:
The subtlety here is that the leader is doing two things:
1. Explicitly determining the start of a basic block
2. Implicitly delimiting the _end_ of any (linearly) preceding block.

For use (1), the addr of the bytes after an unconditional jump may be incorrect in some corner cases. In particular, if they are not actually code. For (2), we want the non-fallthrough next instruction as a leader, _even if_ it is unreachable. If we didn't treat it as a leader, then the second sweep, which takes us from leaders to blocks, would give us "blocks" with unconditional jumps in the middle of the instruction stream!

I think this distinction is muddled by algorithms which (in an abstracted setting) assume non-mixing of code and data.